clientv3/integration/kv_test.go

```diff
@@ -288,14 +288,8 @@ func TestKVGetErrConnClosed(t *testing.T) {
 	cli := clus.Client(0)
 	kv := clientv3.NewKV(cli)
 
-	closed, donec := make(chan struct{}), make(chan struct{})
+	donec := make(chan struct{})
 	go func() {
-		select {
-		case <-time.After(3 * time.Second):
-			t.Fatal("cli.Close took too long")
-		case <-closed:
-		}
-
```

Comment:
Or how about just move `kv := clientv3.NewKV(cli)` below here?

```go
package main

import (
	"fmt"
	"log"
	"time"

	"golang.org/x/net/context"

	"github.com/coreos/etcd/clientv3"
)

var (
	dialTimeout    = 5 * time.Second
	requestTimeout = 1 * time.Second
	endpoints      = []string{"localhost:2379"}
)

func main() {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   endpoints,
		DialTimeout: dialTimeout,
	})
	if err != nil {
		log.Fatal(err)
	}

	closed, donec := make(chan struct{}), make(chan struct{})
	// donec := make(chan struct{})
	go func() {
		select {
		case <-time.After(3 * time.Second):
			log.Fatal("cli.Close took too long")
		case <-closed:
		}

		kv := clientv3.NewKV(cli)

		// ctx, cancel := context.WithTimeout(cli.Ctx(), 3*time.Second) // cli.Close cancels this context, and will panic
		ctx, cancel := context.WithTimeout(context.TODO(), 3*time.Second) // cli.Close does not cancel this context, and still panic

		fmt.Println("kv.Get 1", ctx.Err())
		_, err := kv.Get(ctx, "test")
		fmt.Println("kv.Get 2")

		fmt.Println("cancel")
		cancel()
		fmt.Println("canceled")

		if err != nil {
			log.Fatal(err)
		}
		close(donec)
	}()

	if err := cli.Close(); err != nil {
		log.Fatal(err)
	}
	close(closed)

	select {
	case <-time.After(3 * time.Second):
		log.Fatal("kv.Get took too long")
	case <-donec:
	}
	fmt.Println("DONE")
}
```

But below doesn't panic without the previous patch:

```go
package main

import (
	"fmt"
	"log"
	"time"

	"golang.org/x/net/context"

	"github.com/coreos/etcd/clientv3"
)

var (
	dialTimeout    = 5 * time.Second
	requestTimeout = 1 * time.Second
	endpoints      = []string{"localhost:2379"}
)

func main() {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   endpoints,
		DialTimeout: dialTimeout,
	})
	if err != nil {
		log.Fatal(err)
	}

	kv := clientv3.NewKV(cli)

	// closed, donec := make(chan struct{}), make(chan struct{})
	donec := make(chan struct{})
	go func() {
		// select {
		// case <-time.After(3 * time.Second):
		// 	log.Fatal("cli.Close took too long")
		// case <-closed:
		// }

		// kv := clientv3.NewKV(cli)

		// ctx, cancel := context.WithTimeout(cli.Ctx(), 3*time.Second) // cli.Close cancels this context, and will panic
		ctx, cancel := context.WithTimeout(context.TODO(), 3*time.Second) // cli.Close does not cancel this context, and still panic

		fmt.Println("kv.Get 1", ctx.Err())
		_, err := kv.Get(ctx, "test")
		fmt.Println("kv.Get 2")

		fmt.Println("cancel")
		cancel()
		fmt.Println("canceled")

		if err != nil {
			log.Fatal(err)
		}
		close(donec)
	}()

	if err := cli.Close(); err != nil {
		log.Fatal(err)
	}
	// close(closed)

	select {
	case <-time.After(3 * time.Second):
		log.Fatal("kv.Get took too long")
	case <-donec:
	}
	fmt.Println("DONE")
}

```


/cc @heyitsanthony